### PR TITLE
Include tumor-aware results when results have been rolled-up (sample or library)

### PR DESCRIPTION
**Pull request by yfarjoun**
- Allows for a output that includes tumor-aware crosscheck in all "roll-up" levels (read-group, library, sample, and file)
- Keeps the option to output a matrix (regular LOD only) output
- Adds option to crosscheck vcfs (and a mixture of vcfs and bams)
- Renames the Crosscheck tool to CrosscheckFingerprints (dropping "ReadGroup"), though keeping a deprecated version of the tool with the original name wrapping the renamed tool.
- Added a tool to cluster (non-matrix) output according to high LOD scores connections.

### CODE REVIEW

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 1]
```diff
@@ -132,69 +132,72 @@ public static void main(final String[] args) {
...
+        final SortedSet<String> samples = (SortedSet<String>) sampleFps.keySet();
+
+        // Print header row
+        out.println(type + "1\t" + type + "2\tLOD_SCORE\tLOD_SCORE_TUMOR_NORMAL\tLOD_SCORE_NORMAL_TUMOR");
+
+        for (final String sample : samples) {
+            final Fingerprint fp = sampleFps.get(sample);
+            for (final String otherSample : samples) {
+                final MatchResults results = FingerprintChecker.calculateMatchResults(fp, sampleFps.get(otherSample), GENOTYPING_ERROR_RATE, LOSS_OF_HET_RATE);
+                final List<String> elements = new ArrayList<>(4);
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: (5)?

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 2]
```diff
@@ -132,69 +132,72 @@ public static void main(final String[] args) {
...
+                elements.add(formatUtil.format(results.getLOD()));
+                elements.add(formatUtil.format(results.getLodTN()));
+                elements.add(formatUtil.format(results.getLodNT()));
+                out.println(String.join("\t", elements));
+            }
+        }
+    }
+
+
+        /**
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: Shouldn't this block be further left?

**Comment by gbggrant** [comment]
👍

**Comment by yfarjoun** [comment]
I had to rebase onto your changes @gbggrant and I took the opportunity to re-write everything. 

Can you please re-review please?

File: src/main/java/picard/fingerprint/CrosscheckMetric.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2010 The Broad Institute
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckMetric.java]: 2010?

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 4]
```diff
@@ -26,104 +26,100 @@
...
+
+import static java.util.stream.Collectors.toSet;
+import static picard.fingerprint.CrosscheckMetric.FingerprintResult.*;
+import static picard.fingerprint.CrosscheckMetric.FingerprintResult;
 
 /**
- * Program to check that all read groups within the set of BAM files appear to come from the same
- * individual.
+ * Program to check that all (read-)groups within the set of BAM files appear to come from the same
+ * individual. Can br used to cross-check libraries or samples too.
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: br => 'be'

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 5]
```diff
@@ -26,104 +26,100 @@
...
-            doc="If any two read groups match with a LOD score lower than the threshold the program will exit " +
-                "with a non-zero code to indicate error. 0 means equal probability the read groups match vs. " +
-                "come from different individuals, negative numbers mean N logs more likely that the read groups " +
-                "are from different individuals and positive numbers mean N logs more likely that the read groups " +
-                "are from the sample individual.")
+    @Option(shortName = "LOD",
+            doc = "If any two groups (with the same sample name) match with a LOD score lower than the threshold " +
+                    "the program will exit with a non-zero code to indicate error." +
+                    " Program will also exit with an error if it finds two groups with different sample name that" +
+                    "match with a LOD score greater than -LOD_THRESHOLD." +
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: Need a space at the beginning of this string (or at the end of the string on the line above)

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 6]
```diff
@@ -26,104 +26,100 @@
...
+                    "\n\n" +
+                    "LOD score 0 means equal likelihood" +
+                    "that the groups match vs. come from different individuals, negative LOD scores mean N logs more likely " +
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: Need a space at the beginning of this string (or at the end of the string on the line above)

File: src/main/java/picard/fingerprint/FingerprintIdDetails.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2010 The Broad Institute
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/FingerprintIdDetails.java]: 2010?

File: src/main/java/picard/fingerprint/FingerprintIdDetails.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        FingerprintIdDetails that = (FingerprintIdDetails) o;
+
+        if (platformUnit != null ? !platformUnit.equals(that.platformUnit) : that.platformUnit != null) return false;
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/FingerprintIdDetails.java]: this will NPE if that.platformUnit is null

Comment (gbggrant) [inline src/main/java/picard/fingerprint/FingerprintIdDetails.java]: Same for the next five lines

File: src/test/java/picard/analysis/CollectGcBiasMetricsTest.java [hunk 9]
```diff
@@ -234,11 +234,13 @@ public void runWindowsComparisonTest() throws IOException {
     public File build (final List<SAMRecordSetBuilder> setBuilder, final File unsortedSam, final SAMFileHeader header) throws IOException {
         final File sortedSam = File.createTempFile("CollectGcBias", ".bam", TEST_DIR);
         sortedSam.deleteOnExit();
+        final File sortedSamIdx = new File(TEST_DIR,sortedSam.getName()+".idx");
```
Comment (gbggrant) [inline src/test/java/picard/analysis/CollectGcBiasMetricsTest.java]: spacing

**Comment by gbggrant** [comment]
👍

**Comment by yfarjoun** [comment]
need to add matrix output.

**Comment by yfarjoun** [comment]
sorry. had to rebase....do you remember this PR @gbggrant ? Can you take another look?

**Comment by yfarjoun** [comment]
bump. @gbggrant . please let me know if you will not be able to review this.

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 10]
```diff
@@ -70,7 +72,23 @@
             doc = "Optional output file to write metrics to. Default is to write to stdout.")
     public File OUTPUT;
 
-    @Option(shortName = "H", doc = "The file of haplotype data to use to fingerprint")
+    @Option(shortName = "MO", optional = true,
+            doc = "Optional output file to write matrix of LOD scores to. This is less informative than the metrics output and only contains \"headling\" LOD score." +
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: "heading" - not headling

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 11]
```diff
@@ -70,7 +72,23 @@
...
+            "followed by a tab-separated header line and lines describing SNPs and Haplotype blocks. For example :" +
+            "#CHROMOSOME    POSITION    NAME        MAJOR_ALLELE    MINOR_ALLELE    MAF         ANCHOR_SNP  PANELS" +
+            "1              29478419    rs2503005   C               T               0.483791    rs2230678   panel1" +
+            "1              29497820    rs1994859   C               T               0.483791    rs2230678   panel2" +
+            "1              29500995    rs2486204   T               C               0.484167    rs2230678         " +
+            "1              29509603    rs2428556   C               T               0.728976    rs2230679   panel1" +
+            "" +
+            "NOTE: SNPs listed with the same ANCHOR_SNP will be in the same haplotype. In case of discrepancy between the MAFs within a block, " +
+            "the MAF of the first (smallest genomic position) SNP in the block is considered the MAF of the block." +
+            "NOTE: the PANEL field is optional (as a value, not in the header)")
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: PANEL or PANELS (it's shown as PANELS in the example header line)

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 12]
```diff
@@ -165,6 +192,38 @@ protected int doWork() {
...
+        final NumberFormat format =  NumberFormat.getInstance();
+        format.setMaximumFractionDigits(4);
+
+        try (final OutputStream stream = new FileOutputStream(MATRIX_OUTPUT);
+             final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(stream))) {
+
+            // write the type by which the roll-up happened in the top left corner of the matrix
+            writer.write(CROSSCHECK_BY.name());
+
+            // write the nams of the keys as the first row
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: "names"

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 13]
```diff
@@ -228,21 +287,29 @@ private int crossCheckGrouped(final Map<FingerprintIdDetails, Fingerprint> finge
 
                         }));
 
-        return crossCheckFingerprints(fingerprintsByGroup, type, metrics);
+        if (MATRIX_OUTPUT!=null) {
+           crosscheckMatrix = new double [fingerprintsByGroup.size()][];
+            for(int i =0 ; i< fingerprintsByGroup.size(); i++){
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: argh. Such annoying spacing!

File: src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java [hunk 14]
```diff
@@ -228,21 +287,29 @@ private int crossCheckGrouped(final Map<FingerprintIdDetails, Fingerprint> finge
...
+        }
+        return crossCheckFingerprints(fingerprintsByGroup, type, metrics, by);
     }
 
     /**
      * Method that pairwise checks every pair of read groups and reports a LOD score for the two read groups
      * coming from the same sample.
      */
-    private int crossCheckFingerprints(final Map<FingerprintIdDetails, Fingerprint> fingerprints, final CrosscheckMetric.DataType type, final List<CrosscheckMetric> metrics) {
+    private int crossCheckFingerprints(final Map<FingerprintIdDetails, Fingerprint> fingerprints, final CrosscheckMetric.DataType type, final List<CrosscheckMetric> metrics, Function<FingerprintIdDetails, String> by) {
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java]: could 'by' be final?  And, 'by' - really?

File: src/main/java/picard/fingerprint/CrosscheckFingerprints.java [hunk 15]
```diff
@@ -51,15 +51,25 @@
...
+ * Operates on bams/sams and vcfs.
  *
  * @author Tim Fennell
  * @author Yossi Farjoun
  */
 @CommandLineProgramProperties(
-        usage = "Checks if all (read-)groups within a set of files appear to come from the same individual.",
-        usageShort = "Checks if all (read-)groups appear to come from the same individual.",
+        usage = "Checks if all fingerprints within a set of files appear to come from the same individual." +
+                "The fingerprints are calculated initialy at the readgroup level (if present) but can be" +
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckFingerprints.java]: "initially"

File: src/main/java/picard/fingerprint/CrosscheckFingerprints.java [hunk 16]
```diff
@@ -51,15 +51,25 @@
...
+                "\"rolled-up\" by libarary, sample or file, to increase power and provide results at the " +
+                "desired resolution. Regular output is in a \"Moltenized\" format, one row per comparison, " +
+                "In this format the output will include the LOD score and also tumor-aware LOD score which can " +
+                "help assess identity even in the presence of a severe LOH sample with high purity." +
+                "A matrix output is also availalble to facilitate visual inspection of crosscheck results." +
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckFingerprints.java]: available

File: src/main/java/picard/fingerprint/CrosscheckFingerprints.java [hunk 17]
```diff
@@ -51,15 +51,25 @@
...
+                "\n" +
+                "A separate CLP, ClusterCrosscheckMetrics, can cluster the results as a connected graph" +
```
Comment (gbggrant) [inline src/main/java/picard/fingerprint/CrosscheckFingerprints.java]: Need a trailing space on this line, and several others (above) in this text block.

**Comment by sooheelee** [comment]
Let me know if this needs documentation.

**Comment by yfarjoun** [comment]
comment from @ktibbett: make old CLP argument not change (using a wrapper)

**Comment by yfarjoun** [comment]
@gbggrant so...since your review in April, I added a few features...could you do another look-see?

**Comment by yfarjoun** [comment]
back to you @gbggrant 

The intention was that the original CrossCheckReadgroupFingerprints will not change its API. In particular, I didn't (intend to) change any of its tests. So, I'm not sure what you mean by your comment about "zamboni". 

Could you see if you agree that current behavior will not change?

**Comment by gbggrant** [comment]
@yfarjoun 
I had thought that the inputs to CrossCheckReadgroupFingerprints had changed - but looking at it again I see that they haven't.

**Pull request merged**